File: diffusers_3d/ops/ball_query.py
import torch


@torch.no_grad()
def ball_query(
    points: torch.Tensor, queries: torch.Tensor, radius: float, max_samples_per_query: int
) -> torch.Tensor:
    points = points.contiguous()
    queries = queries.contiguous()

    return torch.ops.diffusers_3d.ball_query(
        points, queries, radius, max_samples_per_query
    )


def main():
    points = torch.randn(3, 1024, 3, dtype=torch.float32, device="cuda")
    idxs = ball_query(points, points, 0.1, 50)
    print(idxs.shape, idxs.min(), idxs.max())


if __name__ == "__main__":
    main()

File: src/diffusers_3d/ball_query.cpp
#include <torch/library.h>
#include "diffusers_3d/ball_query.h"

namespace diffusers_3d::ball_query {

at::Tensor ball_query(
    const at::Tensor& points,
    const at::Tensor& queries,
    double radius,
    int64_t max_samples_per_query) {
  TORCH_CHECK(points.is_contiguous(), "points must be contiguous");
  TORCH_CHECK(queries.is_contiguous(), "queries must be contiguous");

  TORCH_CHECK(points.dim() == 3, "points must be a 3D tensor");
  TORCH_CHECK(queries.dim() == 3, "queries must be a 3D tensor");

  static auto op = c10::Dispatcher::singleton()
                       .findSchemaOrThrow("diffusers_3d::ball_query", "")
                       .typed<decltype(ball_query)>();
  return op.call(points, queries, radius, max_samples_per_query);
}

TORCH_LIBRARY_FRAGMENT(diffusers_3d, m) {
  m.def(TORCH_SELECTIVE_SCHEMA(
      "diffusers_3d::ball_query(Tensor points, Tensor queries, "
      "float radius, int max_samples_per_query) -> Tensor"));
}

} // namespace diffusers_3d::ball_query
